modules/grpc/otel/otel-protobuf-formatter.cpp

```diff
@@ -329,6 +347,57 @@ ProtobufFormatter::get_metadata(LogMessage *msg, Resource &resource, std::string
          get_scope_and_schema_url(msg, scope, scope_schema_url);
 }
 
+bool
+ProtobufFormatter::format(LogMessage *msg, LogRecord &log_record)
+{
+  gssize len;
+  const gchar *value;
+
+  value = _get_protobuf(msg, ".otel_raw.log", &len);
+  if (value)
+    return log_record.ParseFromArray(value, len);
+
+  log_record.set_time_unix_nano(_get_uint64(msg, ".otel.log.time_unix_nano"));
+  log_record.set_observed_time_unix_nano(_get_uint64(msg, ".otel.log.observed_time_unix_nano"));
+
+  int32_t severity_number_int = _get_int32(msg, ".otel.log.severity_number");
+  SeverityNumber severity_number = SeverityNumber_IsValid(severity_number_int)
+                                   ? (SeverityNumber) severity_number_int
+                                   : SEVERITY_NUMBER_UNSPECIFIED;
+  log_record.set_severity_number(severity_number);
+
+  value = _get_string(msg, ".otel.log.severity_text", &len);
+  log_record.set_severity_text(value, len);
+
+  _get_and_set_AnyValue(msg, ".otel.log.body", log_record.mutable_body());
+
+  get_and_set_repeated_KeyValues(msg, ".otel.log.attributes.", log_record.mutable_attributes());
+
+  log_record.set_dropped_attributes_count(_get_uint32(msg, ".otel.log.dropped_attributes_count"));
+
+  log_record.set_flags(_get_uint32(msg, ".otel.log.flags"));
+
+  value = _get_bytes(msg, ".otel.log.trace_id", &len);
+  log_record.set_trace_id(value, len);
+
+  value = _get_bytes(msg, ".otel.log.span_id", &len);
+  log_record.set_span_id(value, len);
+
+  return true;
+}
+
+bool
+ProtobufFormatter::format_fallback(LogMessage *msg, LogRecord &log_record)
+{
+  log_record.set_time_unix_nano(msg->timestamps[LM_TS_STAMP].ut_sec * 1000000000 +
+                                msg->timestamps[LM_TS_STAMP].ut_usec * 1000);
+  log_record.set_observed_time_unix_nano(msg->timestamps[LM_TS_RECVD].ut_sec * 1000000000 +
+                                         msg->timestamps[LM_TS_RECVD].ut_usec * 1000);
+  log_record.set_severity_number(_get_log_msg_severity_number(msg));
+  _get_and_set_AnyValue(msg, "MESSAGE", log_record.mutable_body());
+
+  return true;
```

Comment:
Fallback should never fail in my opinion

Reply:
Thanks, fixed!